incubator/hnc/internal/validators/hierarchy.go

```diff
@@ -210,9 +212,55 @@ func (v *Hierarchy) checkParent(ns, curParent, newParent *forest.Namespace) admi
 		return deny(metav1.StatusReasonConflict, "Illegal parent: "+reason)
 	}
 
+	// Prevent overwriting source objects in the descendants after the hierarchy change.
+	if yes, co := v.hasConflictObject(newParent, ns); yes {
+		coStr := strings.Join(co, "\n  * ")
+		msg := fmt.Sprintf("\nCannot update hierarchy because it would overwrite the following object(s):\n  * %s\nTo fix this, please rename or remove the conflicting objects first.", coStr)
+		return deny(metav1.StatusReasonConflict, msg)
+	}
+
 	return allow("")
 }
 
+// hasConflictObject returns true and a list of objects if there's a conflict.
+func (v *Hierarchy) hasConflictObject(pns, cns *forest.Namespace) (bool, []string) {
+	// Get all types with 'Propagate' mode.
+	ts := []forest.TypeSyncer{}
+	for _, t := range v.Forest.GetTypeSyncers() {
+		if t.GetMode() == api.Propagate {
+			ts = append(ts, t)
+		}
+	}
+
+	// Get all the original objects in the new ancestors (including the parent).
+	existsInAns := make(map[schema.GroupVersionKind]map[string]bool)
+	for _, t := range ts {
+		gvk := t.GetGVK()
+		existsInAns[gvk] = make(map[string]bool)
+		os := append(pns.GetPropagatedObjects(gvk), pns.GetOriginalObjects(gvk)...)
```

Comment:
This line probably deserves a comment. Also, once we have exceptions, we'll need something smarter than this, and you can comment that too (cc @GinnyJI ).

Reply:
Added it as a new function in forest with `TODO`:
```
// GetPropagatingObjects returns all the source objects to be propagated into the
// descendants of this namespace.
// TODO update this function to reflect the changes introduced by the future HNC
//  exceptions implementation.
func (ns *Namespace) GetPropagatingObjects(gvk schema.GroupVersionKind) []*unstructured.Unstructured {
	return append(ns.GetPropagatedObjects(gvk), ns.GetOriginalObjects(gvk)...)
}
```